core/src/main/java/nl/nn/adapterframework/jdbc/JdbcListener.java

```diff
@@ -245,99 +246,113 @@ private String getValueOrDefaultIfColumnDoesNotExistInTable(ResultSet rs, String
-		String mid;
 		String cid;
-		if (rawMessage instanceof IMessageWrapper) {
-			mid = ((IMessageWrapper)rawMessage).getId();
-			Map<String,Object> mwContext = ((IMessageWrapper)rawMessage).getContext();
-			cid = (String)mwContext.get(CORRELATION_ID_KEY);
-			key = (String)mwContext.get(STORAGE_KEY_KEY);
-			if (StringUtils.isEmpty(key)) {
-				key = mid; // backward compatibility
+		String mid;
+		if (rawMessage.getId() != null) {
+			mid = rawMessage.getId();
+		} else if (rawMessage instanceof MessageWrapper) {
+			try {
+				mid = ((MessageWrapper<M>) rawMessage).getMessage().asString();
+			} catch (IOException e) {
+				throw new ListenerException(e);
 			}
+		} else if (rawMessage.getRawMessage() != null) {
+			mid = rawMessage.getRawMessage().toString();
 		} else {
-			key = (String)rawMessage;
-			mid = key;
-			cid = mid;
+			mid = null;
 		}
-		if (threadContext!=null) {
-			PipeLineSession.setListenerParameters(threadContext, mid, cid, null, null);
-			threadContext.put(STORAGE_KEY_KEY, key);
+		Map<String,Object> mwContext = rawMessage.getContext();
+		cid = (String)mwContext.getOrDefault(CORRELATION_ID_KEY, mid);
+		key = (String)mwContext.getOrDefault(STORAGE_KEY_KEY, mid);
+		if (StringUtils.isEmpty(key)) {
+			key = mid; // backward compatibility
 		}
+		updateThreadContextWithIds(threadContext, key, cid, mid);
 		return mid;
 	}
 
-	protected String getKeyFromRawMessage(M rawMessage) throws ListenerException {
+	@Override
+	public String getIdFromRawMessage(M rawMessage, Map<String, Object> threadContext) throws ListenerException {
+		String mid = rawMessage != null ? rawMessage.toString() : null;
+		updateThreadContextWithIds(threadContext, mid, mid, mid);
+		return mid;
+	}
+
+	private static void updateThreadContextWithIds(Map<String, Object> threadContext, String key, String cid, String mid) {
+		if (threadContext != null) {
+			PipeLineSession.updateListenerParameters(threadContext, mid, cid, null, null);
+			threadContext.put(STORAGE_KEY_KEY, key);
+		}
+	}
+
+	protected String getKeyFromRawMessage(RawMessageWrapper<M> rawMessage) throws ListenerException {
 		Map<String,Object> context = new HashMap<>();
-		getIdFromRawMessage(rawMessage, context); // populate context with storage key
+		getIdFromRawMessageWrapper(rawMessage, context); // populate context with storage key
 		return (String)context.get(STORAGE_KEY_KEY);
 	}
 
 	@Override
-	public Message extractMessage(M rawMessage, Map<String,Object> threadContext) throws ListenerException {
+	public Message extractMessage(RawMessageWrapper<M> rawMessage, Map<String,Object> threadContext) throws ListenerException {
 		Message message;
-		if (rawMessage instanceof IMessageWrapper) {
-			message = ((IMessageWrapper)rawMessage).getMessage();
+		if (rawMessage instanceof MessageWrapper) {
+			message = ((MessageWrapper)rawMessage).getMessage();
 		} else {
-			message = Message.asMessage(rawMessage);
+			message = Message.asMessage(rawMessage.getRawMessage());
```

Comment:
Je zou voor het gemakt een transient override kunnen maken die getMessage() { return Message.asMessage(getRawMessage()); } doet.

Reply:
Eerst wel gedaan maar toch niet. Want de listener moet eigenlijk de `Message` maken van de raw message data. Dus dat gemak, is me eigenlijk iets te gemakkelijk want het dwingt je niet te kijken of je wel het juiste ding doet op de juiste plaats.